auto_ptr usage causes noisy compilation

As described in https://github.com/glotzerlab/hoomd-blue/issues/430, the compilation of libgetar currently gives many warnings like the following:

```c++
In file included from src/GTAR.cpp:4:0:
src/GTAR.hpp: In member function 'std::auto_ptr<T> gtar_pymodule::gtar::GTAR::readUniform(const string&)':
src/GTAR.hpp:260:25: warning: 'template<class> class std::auto_ptr' is deprecated [-Wdeprecated-declarations]
             return std::auto_ptr<T>(new T(*(T*) bytes.get()));
```

--- a/src/GTAR.hpp
+++ b/src/GTAR.hpp
@@ -119,7 +119,7 @@ namespace gtar{
         /// Read a uniform binary property to the specified location,
         /// converting from little endian if necessary.
         template<typename T>
-        std::auto_ptr<T> readUniform(const std::string &path);
+        std::unique_ptr<T> readUniform(const std::string &path);
         /// Read a bytestring from the specified location
         SharedArray<char> readBytes(const std::string &path);
 
@@ -162,7 +162,7 @@ namespace gtar{
         void insertRecord(const std::string &path);
 
         /// The archive abstraction object we'll use
-        std::auto_ptr<Archive> m_archive;
+        std::unique_ptr<Archive> m_archive;
 
         /// Cached record objects
         std::map<Record, indexSet> m_records;
@@ -250,16 +250,16 @@ namespace gtar{
     }
 
     template<typename T>
-    std::auto_ptr<T> GTAR::readUniform(const std::string &path)
+    std::unique_ptr<T> GTAR::readUniform(const std::string &path)
     {
         SharedArray<char> bytes(m_archive->read(path));
 
         maybeSwapEndian<T>((T*) bytes.get(), bytes.size());
 
         if(bytes.size())
-            return std::auto_ptr<T>(new T(*(T*) bytes.get()));
+            return std::unique_ptr<T>(new T(*(T*) bytes.get()));
         else
-            return std::auto_ptr<T>();
+            return std::unique_ptr<T>();
     }
 
 }

--- a/src/GTAR.hpp
+++ b/src/GTAR.hpp
@@ -8,6 +8,13 @@
 #include <string>
 #include <vector>
 
+// if C++11 or greater, use unique_ptr
+#if __cplusplus > 199711L
+#define gtar_unique_ptr std::unique_ptr
+#else
+#define gtar_unique_ptr std::auto_ptr
+#endif
+
 #include "Archive.hpp"
 #include "DirArchive.hpp"
 #include "SqliteArchive.hpp"
@@ -119,7 +126,7 @@ namespace gtar{
         /// Read a uniform binary property to the specified location,
         /// converting from little endian if necessary.
         template<typename T>
-        std::unique_ptr<T> readUniform(const std::string &path);
+        SharedPtr<T> readUniform(const std::string &path);
         /// Read a bytestring from the specified location
         SharedArray<char> readBytes(const std::string &path);
 
@@ -162,7 +169,7 @@ namespace gtar{
         void insertRecord(const std::string &path);
 
         /// The archive abstraction object we'll use
-        std::unique_ptr<Archive> m_archive;
+        gtar_unique_ptr<Archive> m_archive;
 
         /// Cached record objects
         std::map<Record, indexSet> m_records;
@@ -250,16 +257,16 @@ namespace gtar{
     }
 
     template<typename T>
-    std::unique_ptr<T> GTAR::readUniform(const std::string &path)
+    SharedPtr<T> GTAR::readUniform(const std::string &path)
     {
         SharedArray<char> bytes(m_archive->read(path));
 
         maybeSwapEndian<T>((T*) bytes.get(), bytes.size());
 
         if(bytes.size())
-            return std::unique_ptr<T>(new T(*(T*) bytes.get()));
+            return SharedPtr<T>((SharedArray<T>&) bytes);
         else
-            return std::unique_ptr<T>();
+            return SharedPtr<T>();
     }
 
 }

--- a/src/SharedArray.hpp
+++ b/src/SharedArray.hpp
@@ -13,6 +13,7 @@ namespace GTAR_NAMESPACE_PARENT{
 namespace gtar{
 
 template<typename T> class SharedArray;
+template<typename T> class SharedPtr;
 
 /// Shim for the SharedArray class. Wraps the reference counting and
 /// pointer storage for a SharedArray.
@@ -61,6 +62,7 @@ private:
 template<typename T>
 class SharedArray
 {
+    friend class SharedPtr<T>;
 public:
     typedef T* iterator;
 
@@ -78,6 +80,15 @@ public:
     /// Copy constructor: make this object point to the same array as
     /// rhs, increasing the reference count if necessary
     SharedArray(const SharedArray<T> &rhs):
+        m_shim(rhs.m_shim)
+    {
+        if(m_shim)
+            m_shim->increment();
+    }
+
+    /// Initialize from SharedPtr: make this object point to the same
+    /// shim as rhs, increasing the reference count if necessary
+    SharedArray(const SharedPtr<T> &rhs):
         m_shim(rhs.m_shim)
     {
         if(m_shim)
@@ -192,6 +203,57 @@ private:
     /// Our pointer to the shim, which holds the array pointer and
     /// reference count
     SharedArrayShim<T> *m_shim;
+};
+
+/// Exposes a pointer-like interface to a single element for a
+/// SharedArray's contents. This is only a shim; elements should still
+/// be allocated using new[], for example.
+template<typename T>
+class SharedPtr: public SharedArray<T>
+{
+public:
+    /// Default constructor. Allocates nothing.
+    SharedPtr():
+        SharedArray<T>()
+    {}
+
+    /// Target constructor: allocates a new SharedArrayShim for the
+    /// given pointer and takes ownership of it. target will be
+    /// deallocated using delete[] after all references to it have
+    /// been removed.
+    SharedPtr(T *target):
+        SharedArray<T>(target, 1)
+    {}
+
+    /// Copy constructor: make this object point to the same array as
+    /// rhs, increasing the reference count if necessary
+    SharedPtr(const SharedPtr<T> &rhs):
+        SharedArray<T>(rhs)
+    {}
+
+    /// Initialize from SharedArray: make this object point to the
+    /// same shim as rhs, increasing the reference count if necessary
+    SharedPtr(const SharedArray<T> &rhs):
+        SharedArray<T>(rhs)
+    {}
+
+    T &operator*()
+    {
+        return *(this->get());
+    }
+    const T &operator*() const
+    {
+        return *(this->get());
+    }
+
+    T* operator->()
+    {
+        return this->get();
+    }
+    const T* operator->() const
+    {
+        return this->get();
+    }
 };
 
 }

--- a/src/ZipArchive.cpp
+++ b/src/ZipArchive.cpp
@@ -14,7 +14,7 @@ namespace GTAR_NAMESPACE_PARENT{
 
 namespace gtar{
 
-    using std::auto_ptr;
+    using std::unique_ptr;
     using std::runtime_error;
     using std::string;
     using std::stringstream;

--- a/src/ZipArchive.cpp
+++ b/src/ZipArchive.cpp
@@ -14,7 +14,6 @@ namespace GTAR_NAMESPACE_PARENT{
 
 namespace gtar{
 
-    using std::unique_ptr;
     using std::runtime_error;
     using std::string;
     using std::stringstream;

--- a/test/test_GTAR.cpp
+++ b/test/test_GTAR.cpp
@@ -19,7 +19,7 @@ void runTests(int &result, string suffix)
         arch.close();
 
         GTAR readArch("test" + suffix, Read);
-        unique_ptr<int> readUniform(readArch.readUniform<int>("test.i64.uni"));
+        SharedPtr<int> readUniform(readArch.readUniform<int>("test.i64.uni"));
 
         if(!readUniform.get())
         {

--- a/test/test_GTAR.cpp
+++ b/test/test_GTAR.cpp
@@ -19,7 +19,7 @@ void runTests(int &result, string suffix)
         arch.close();
 
         GTAR readArch("test" + suffix, Read);
-        auto_ptr<int> readUniform(readArch.readUniform<int>("test.i64.uni"));
+        unique_ptr<int> readUniform(readArch.readUniform<int>("test.i64.uni"));
 
         if(!readUniform.get())
         {

